File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        combinationSumHelper(nums, target, nums.length-1, curr, ans);
        return ans;
    }
    public void combinationSumHelper(int[] nums, int target, int n, List<Integer> current,List<List<Integer>> ans){
        if(target < 0 || n < 0){
            return;
        }
        if(target == 0){
            ans.add(new ArrayList<Integer>(current));
            return;
        }
        current.add(nums[n]);
        combinationSumHelper(nums,target-nums[n],n,current,ans);
        current.remove(current.size()-1);
        combinationSumHelper(nums,target,n-1,current,ans);
    }
}

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def helper(self, nums, current_set, ans):
        if len(current_set) == len(nums):
            ans.append(current_set[:])
            return
        for i in nums:
            if i not in current_set:
                current_set.append(i)
                self.helper(nums, current_set, ans)
                current_set.pop()

    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.helper(nums, [], ans)
        return ans
        